### PR TITLE
Fix e2e test failure screenshots not capturing at the right time

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

Continued from #26664. Fix the screenshots not being captured at the right time.

Specifically, if the tests have `afterEach` or `afterAll` defined, the screenshots will be captured after these hooks are run. This PR uses `jest-circus` as the test runner and extends `jest-environment-puppeteer` to run the capturing scripts in the `test_fn_failure` event.

Not sure why the lock file has so many changes though.

## How has this been tested?

Added some intentionally failing tests in https://github.com/WordPress/gutenberg/runs/2358561781 to demo the result. See the failing tests for the stored artifacts. The screenshots captured should correctly point to when it failed.

```sh
npm run test-e2e -- packages/e2e-tests/specs/widgets/adding-widgets.test.js
```

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by kevin940726** [comment]
Somehow when using `jest-circus`, there will be an unexpected error when running `multi-entity-editing.test.js`.

```
Test suite failed to run

Expected an Error, but "" was thrown
```

I've narrowed it down to this following call:

```js
await createNewPost({
  postType: "wp_template",
  title: templateName
});
```

Specifically the `wp_template` post type.

I have no idea why it's broken, and I don't have time to investigate it further now, might come back to it later.

**Comment by kevin940726** [comment]
Found the root cause why it was breaking and fixed it.

```
Test suite failed to run

Expected an Error, but "" was thrown
```

The above error shown in the `multi-entity-editing` tests are coming from an unhandled exception in the "creating wp_template post" page. You can see the same error being thrown in a more descriptive way when visiting http://localhost:8889/wp-admin/post-new.php?post_type=wp_template&post_title=Custom+Template.

```
asyncToGenerator.js:6
Uncaught (in promise)
{code: "rest_template_not_found", message: "No templates exist with that id.", data: { "status": 404 }}
```

I know very little about the template API so I don't know whether it's expected or not. Either way, it's already happening in trunk just that it didn't throw any error before.

In trunk, we're using the default test runner `jest-jasmine2`. It doesn't throw any error if there's an unexpected exception, instead, it simply log the error: https://github.com/facebook/jest/blob/HEAD/packages/jest-jasmine2/src/jasmine/Env.ts#L250-L251. In this PR though, we're using `jest-circus` which by default throw any unexpected error: https://github.com/facebook/jest/blob/1be8d737abd0e2f30e3314184a0efc372ad6d88f/packages/jest-circus/src/globalErrorHandlers.ts#L13. However, the error message is swallowed in the puppeteer context.

To mimic the same behavior in `jest-jasmine2`, instead of emitting the uncaught exception, I log the error out to prevent the test from failing. Ideally, I think we should fix the root cause though.

**Review by talldan — APPROVED** [review]
I tested locally by adding `expect( 1 ).toBe( 2 )` in a test, and the artifacts folder had the correct screenshot 👍 🚢 

It also made me realise I should delete the artifacts folder now and then 😮

File: packages/scripts/config/jest-e2e.config.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ const path = require( 'path' );
 const { hasBabelConfig } = require( '../utils' );
 
 const jestE2EConfig = {
+	testRunner: 'jest-circus/runner',
```
Comment (gziolo) [inline packages/scripts/config/jest-e2e.config.js:12]: Should we update the test runner for unit tests as well for consistency?

**Review by gziolo — COMMENTED** [review]
Any ideas why the changes to the package lock file are so big?

Nice one. It needs some documentation in the `@wordpress/scripts` package to let people know that a new feature was added and where can expect to see failure screenshots on their disk.

**Comment by kevin940726** [comment]
> Any ideas why the changes to the package lock file are so big?

I think that's because the lock file in trunk was outdated. I believe it has been resolved recently though. A rebase would fix that.

> Nice one. It needs some documentation in the `@wordpress/scripts` package to let people know that a new feature was added and where can expect to see failure screenshots on their disk.

It's not a new feature though, it was being added in #26664. This PR fixes a bug which caused it to not function correctly when there're async `afterAll`/`afterEach`.

File: packages/scripts/config/jest-e2e.config.js (hunk 1 above)
Comment (kevin940726) [reply packages/scripts/config/jest-e2e.config.js:12]: Worth a shot!

**Comment by adamziel** [comment]
Good catch!

**Comment by gziolo** [comment]
> > Nice one. It needs some documentation in the `@wordpress/scripts` package to let people know that a new feature was added and where can expect to see failure screenshots on their disk.
> 
> It's not a new feature though, it was being added in #26664. This PR fixes a bug which caused it to not function correctly when there're async `afterAll`/`afterEach`.

It isn't a new feature in the Gutenberg repository. It's a new feature in the `@wordpress/scripts` package that has its [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/scripts/CHANGELOG.md) file and [README](https://github.com/WordPress/gutenberg/blob/trunk/packages/scripts/README.md) file.

**Comment by kevin940726** [comment]
> It isn't a new feature in the Gutenberg repository. It's a new feature in the `@wordpress/scripts` package that has its [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/scripts/CHANGELOG.md?rgh-link-date=2021-04-19T12%3A55%3A25Z) file and [README](https://github.com/WordPress/gutenberg/blob/trunk/packages/scripts/README.md?rgh-link-date=2021-04-19T12%3A55%3A25Z) file.

Got it, makes sense! I missed that, thanks for the explanation ♥️.

**Comment by kevin940726** [comment]
@draganescu Oops, this is not done yet, still have some notes left to do. I'll open another PR.

File: packages/scripts/config/jest-e2e.config.js (hunk 1 above)
Comment (kevin940726) [reply packages/scripts/config/jest-e2e.config.js:12]: Added in https://github.com/WordPress/gutenberg/pull/31178

**Pull request merged**